src/context/blocked.rs

```diff
@@ -1,18 +1,151 @@
+        }
+    }
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn free_callback_wrapper<T: 'static>(
+    ctx: *mut raw::RedisModuleCtx,
+    privdata: *mut c_void,
+) {
+    let context = Context::new(ctx);
+
+    if privdata.is_null() {
+        context
+            .log_warning("[free_callback_wrapper] Private data is null; this should not happen!");
+        return;
+    }
+
+    let (_, free_cb, data) = *Box::from_raw(
+        privdata
+            as *mut (
+                AuthReplyCallback<T>,
+                Option<FreePrivDataCallback<T>>,
+                Box<T>,
+            ),
+    );
+
+    // free_cb is optional, so we only call free_cb if it exists
+    // If no free_cb exists but the data(priv_data) exists, that's fine -
+    // user might not allocate memory using VM_ALLOC or VM_REALLOC hence
+    // it is not explicitly needed to free using VM_FREE and rust implicit
+    // drop will handle cleanup. If memory was allocated using VM_ALLOC,
+    // the module owners are responsible to provide free_privdata_callback
+    // to avoid memory leaks.
+    if let Some(free_cb) = free_cb {
+        free_cb(&context, *data);
+    }
 }
 
 // We need to be able to send the inner pointer to another thread
-unsafe impl Send for BlockedClient {}
+unsafe impl<T> Send for BlockedClient<T> {}
 
-impl Drop for BlockedClient {
+impl<T> BlockedClient<T> {
+    pub(crate) fn new(inner: *mut raw::RedisModuleBlockedClient) -> Self {
+        Self {
+            inner,
+            auth_reply_callback: None,
+            free_priv_callback: None,
+            private_data: None,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn with_auth_callback(
+        inner: *mut raw::RedisModuleBlockedClient,
+        auth_reply_callback: AuthReplyCallback<T>,
+        free_priv_callback: Option<FreePrivDataCallback<T>>,
+    ) -> Self {
+        Self {
+            inner,
+            auth_reply_callback: Some(auth_reply_callback),
+            free_priv_callback,
+            private_data: None,
+        }
+    }
+
+    pub fn set_blocked_private_data(&mut self, data: T) {
+        self.private_data = Some(Box::new(data));
+    }
+
+    pub fn abort(mut self) -> Result<(), ValkeyError> {
+        unsafe {
+            // Clear references to private_data and callbacks
+            self.private_data = None;
+            self.auth_reply_callback = None;
+            self.free_priv_callback = None;
+
+            if raw::RedisModule_AbortBlock.unwrap()(self.inner) == raw::REDISMODULE_OK as c_int {
+                // Prevent the normal Drop from running
+                self.inner = std::ptr::null_mut();
+                Ok(())
+            } else {
+                Err(ValkeyError::Str("Failed to abort blocked client"))
+            }
+        }
+    }
+}
+
+impl<T> Drop for BlockedClient<T> {
     fn drop(&mut self) {
-        unsafe { raw::RedisModule_UnblockClient.unwrap()(self.inner, ptr::null_mut()) };
+        if !self.inner.is_null() {
```

Comment:
This will leak boxed data for the general blocking case. Can we instead check for private data and only provide a boxed value if private data exists?

Also, we should still call RedisModule_UnblockClient for the generic case. It is just that private data should be null here